File: .claude/hooks/skill-activation-prompt.ts
#!/usr/bin/env bun

interface HookInput {
  session_id: string;
  transcript_path: string;
  cwd: string;
  allowed_tools: string[];
  prompt: string;
}

interface SkillTrigger {
  keywords: string[];
  intent_patterns: string[];
  file_patterns?: string[];
}

interface SkillRule {
  name: string;
  path: string;
  description: string;
  triggers: SkillTrigger;
  priority: "critical" | "high" | "medium" | "low";
}

interface SkillRules {
  version: string;
  skills: SkillRule[];
}

interface MatchedSkill {
  skill: SkillRule;
  matchType: "keyword" | "intent" | "file";
  matchedTerm: string;
}

async function main(): Promise<void> {
  try {
    let inputData = "";
    for await (const chunk of process.stdin) {
      inputData += chunk;
    }

    if (!inputData.trim()) {
      process.exit(0);
    }

    const input: HookInput = JSON.parse(inputData);
    const prompt = input.prompt.toLowerCase();

    // Load skill rules
    const fs = await import("fs/promises");
    const path = await import("path");

    const skillRulesPath = path.join(
      process.env.CLAUDE_PROJECT_DIR || process.cwd(),
      ".claude",
      "skills",
      "skill-rules.json"
    );

    let skillRules: SkillRules;
    try {
      const rulesContent = await fs.readFile(skillRulesPath, "utf-8");
      skillRules = JSON.parse(rulesContent);
    } catch {
      // No skill rules configured, exit silently
      process.exit(0);
    }

    const matchedSkills: MatchedSkill[] = [];
    const projectDir = process.env.CLAUDE_PROJECT_DIR || process.cwd();

    for (const skill of skillRules.skills) {
      // Verify skill file exists before matching
      const skillPath = path.join(projectDir, skill.path);
      try {
        await fs.access(skillPath);
      } catch {
        // Skill file not found, skip
        continue;
      }

      // Check keyword matches
      for (const keyword of skill.triggers.keywords) {
        if (prompt.includes(keyword.toLowerCase())) {
          matchedSkills.push({
            skill,
            matchType: "keyword",
            matchedTerm: keyword,
          });
          break;
        }
      }

      // Check intent pattern matches
      for (const pattern of skill.triggers.intent_patterns) {
        try {
          const regex = new RegExp(pattern, "i");
          if (regex.test(prompt)) {
            // Avoid duplicates
            if (!matchedSkills.find((m) => m.skill.name === skill.name)) {
              matchedSkills.push({
                skill,
                matchType: "intent",
                matchedTerm: pattern,
              });
            }
            break;
          }
        } catch {
          // Invalid regex, skip
        }
      }
    }

    if (matchedSkills.length === 0) {
      process.exit(0);
    }

    // Group by priority
    const priorityOrder = ["critical", "high", "medium", "low"];
    const grouped = new Map<string, MatchedSkill[]>();

    for (const priority of priorityOrder) {
      grouped.set(priority, []);
    }

    for (const match of matchedSkills) {
      const priority = match.skill.priority;
      grouped.get(priority)?.push(match);
    }

    // Format output
    const output: string[] = [];
    output.push("");
    output.push("---");
    output.push("SKILL SUGGESTIONS");
    output.push("---");

    const icons: Record<string, string> = {
      critical: "[!]",
      high: "[*]",
      medium: "[+]",
      low: "[i]",
    };

    for (const priority of priorityOrder) {
      const skills = grouped.get(priority) || [];
      if (skills.length > 0) {
        output.push("");
        output.push(`${icons[priority]} ${priority.toUpperCase()} PRIORITY:`);
        for (const match of skills) {
          output.push(`   - ${match.skill.name}: ${match.skill.description}`);
          output.push(`     Path: ${match.skill.path}`);
          output.push(`     Matched: ${match.matchType} "${match.matchedTerm}"`);
        }
      }
    }

    output.push("");
    output.push(
      "To use a suggested skill, read its SKILL.md file at the provided path with the Read tool, then follow its instructions."
    );
    output.push("---");

    console.log(output.join("\n"));
  } catch (error) {
    // Silently exit on errors to not disrupt workflow
    process.exit(0);
  }
}

main();

File: .claude/hooks/skill_activation_prompt.py
# /// script
# requires-python = ">=3.10"
# ///
"""UserPromptSubmit hook: suggest relevant skills based on the user's prompt.

Loads .claude/skills/skill-rules.json and matches keywords and intent patterns
against the submitted prompt. Groups matches by priority and prints suggestions.
"""

import sys

sys.path.insert(0, str(__import__("pathlib").Path(__file__).parent))

import json
import re
from pathlib import Path

import hook_utils


# ---------------------------------------------------------------------------
# Pure logic functions (testable)
# ---------------------------------------------------------------------------


def load_skill_rules(project_dir: str) -> dict | None:
    """Load skill-rules.json from <project_dir>/.claude/skills/. Returns None on error."""
    rules_path = Path(project_dir) / ".claude" / "skills" / "skill-rules.json"
    try:
        return json.loads(rules_path.read_text())
    except (OSError, json.JSONDecodeError):
        return None


def match_skills(prompt: str, skills: list[dict], project_dir: str) -> list[dict]:
    """Match skills against a prompt.

    Returns a list of match dicts with keys: skill, match_type, matched_term.
    Deduplicates: if a skill already matched via keyword, intent patterns are skipped.
    Skill files that do not exist on disk are skipped entirely.
    """
    prompt_lower = prompt.lower()
    matches: list[dict] = []
    matched_names: set[str] = set()

    for skill in skills:
        skill_path = Path(project_dir) / skill.get("path", "")
        if not skill_path.exists():
            continue

        triggers = skill.get("triggers", {})
        name = skill.get("name", "")

        # Keyword match (case-insensitive substring)
        for keyword in triggers.get("keywords", []):
            if keyword.lower() in prompt_lower:
                matches.append(
                    {"skill": skill, "match_type": "keyword", "matched_term": keyword}
                )
                matched_names.add(name)
                break

        # Intent pattern match (regex, IGNORECASE) — skip if already matched by keyword
        if name not in matched_names:
            for pattern in triggers.get("intent_patterns", []):
                try:
                    if re.search(pattern, prompt, re.IGNORECASE):
                        matches.append(
                            {
                                "skill": skill,
                                "match_type": "intent",
                                "matched_term": pattern,
                            }
                        )
                        matched_names.add(name)
                        break
                except re.error:
                    # Invalid regex — skip
                    pass

    return matches


def format_suggestions(matches: list[dict]) -> str:
    """Format matched skills into a human-readable suggestion block."""
    priority_order = ["critical", "high", "medium", "low"]
    icons: dict[str, str] = {
        "critical": "[!]",
        "high": "[*]",
        "medium": "[+]",
        "low": "[i]",
    }

    grouped: dict[str, list[dict]] = {p: [] for p in priority_order}
    for match in matches:
        priority = match["skill"].get("priority", "low")
        if priority in grouped:
            grouped[priority].append(match)

    lines: list[str] = ["", "---", "SKILL SUGGESTIONS", "---"]

    for priority in priority_order:
        group = grouped[priority]
        if group:
            lines.append("")
            lines.append(f"{icons[priority]} {priority.upper()} PRIORITY:")
            for match in group:
                skill = match["skill"]
                lines.append(f"   - {skill['name']}: {skill['description']}")
                lines.append(f"     Path: {skill['path']}")
                lines.append(
                    f"     Matched: {match['match_type']} \"{match['matched_term']}\""
                )

    lines.append("")
    lines.append(
        "To use a suggested skill, read its SKILL.md file at the provided path with the Read tool, then follow its instructions."
    )
    lines.append("---")

    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------


def main() -> None:
    try:
        data = hook_utils.read_input()
        if not data:
            sys.exit(0)

        prompt: str = data.get("prompt", "")
        if not prompt:
            sys.exit(0)

        project_dir = hook_utils.get_project_dir() or data.get("cwd", "")
        if not project_dir:
            sys.exit(0)

        rules = load_skill_rules(project_dir)
        if not rules:
            sys.exit(0)

        skills = rules.get("skills", [])
        if not skills:
            sys.exit(0)

        matches = match_skills(prompt, skills, project_dir)
        if not matches:
            sys.exit(0)

        print(format_suggestions(matches))
    except Exception:
        # Silently exit on any error — never disrupt workflow
        sys.exit(0)


if __name__ == "__main__":
    main()
